Removed MemoryCell class + code cleanup

psyker156 opened the pull request:
MemoryCell class was hurting the VM initialisation time and limiting the amount of ram the VM could use without causing serious performance issues when launching the VM. MemoryCell class was stripped out. Since the MemoryCell was ultimately  used when writing into memory, the MemoryArray was modified with the addition of the writeMemory method.

Default ram size has been set to 16 MB.

Future work: Instruction fetch unit is a little bit too coupled with the memory array. Maybe it should use the IOController?

cannotparse commented on CapuaEnvironment/MemoryArray/MemoryArray.py:
```diff
@@ -76,12 +75,24 @@ def extractMemory(self, address, length=1):
+        """
+        baseIndex = self._computeArrayIndexFromAddress(address)
+        for i in range(0, length):
+            self._memoryCellArray[baseIndex + i] = values[i]
```
A few style changes here:
- `writeMemory` is taking in `length` and `values`, `values` being a `List` already provides a length using `len`. Allowing length to be specified is redundant and can allow user error. 
- Manual indexing isn't necessary.

I'd suggest the following changes  
```python 3
    def writeMemory(self, address, values):
        """
        This method will overwrite values from a given address
        :param address: int, the address where the overwrite is to happen
        :param values: int list, a list containing values to be writen in memory
        :return: none
        """        
        base = self._computeArrayIndexFromAddress(address) 
        self._memoryCellArray[base:base+len(values)] = values
```
One thing to note is what happens if we go over the range and how it should be handled.

I double checked what `length` is used for in the rest of the function and it's still necessary for other calls( without adding validation to verify inputs such as (0b0010) take less than 4 bytes.)

Note about style: seems like other functions will use simplifications like this where possible. I.E. `extractMemory`

psyker156 replied:
Very good call! Was not aware a slice could be replaced in a single go.

psyker156 replied:
As for simplification and removal of the length param from the various functions, I think this should be part of a different PR in order to keep the code style uniform.

cannotparse commented on CapuaEnvironment/MemoryArray/MemoryArray.py:
```diff
@@ -54,7 +53,7 @@ def __init__(self):
             # hardcoded to all access. This will be adjusted after Capua is "booted up".
-            mc = MemoryCell(permission=0b111, value=MEMORY_CELL_INITIAL_VALUE)
+            mc = MEMORY_CELL_INITIAL_VALUE
             self._memoryCellArray.append(mc)
```
While you are touching this up, you can simplify this to:
```python3
def __init__(self):
         """
         This will initialize the whole memory cell array. No parameters required, we want Capua
         environment to be configurable by the user with minimal code change.
         Configuration for memory cell array are in Configuration.Configuration
         """
         # Create a cell for each memory address available default permission is
         # hardcoded to all access. This will be adjusted after Capua is "booted up".
         self._memoryCellArray = [
              MEMORY_CELL_INITIAL_VALUE
              for x in range(MEMORY_START_AT, MEMORY_END_AT)
         ]
```
This also passes my pep8

psyker156 replied:
Good call

psyker156 commented:
@cannotparse I think you are right, the details overlooked some of the changes. I edited the description. 
Also, good code proposition and comments!
Thanks for the insight!

psyker156 commented:
@cannotparse I added a commit to this PR where I applied pretty much all of you comments - Thanks for these. I also added some testing in that new commit. Previous commit was not complete. Got excited and did PR before it was ready. Thanks for bringing me back in line :P 
Decided to keep it in same PR in a different commit so that people can track how comments were applied and where easily.

cannotparse commented:
Looks good. Good call on adding the function to verify it for you and handling the error.

pg-cse commented:
Thank you for this!
